synapse/http/client.py

```diff
@@ -819,6 +820,125 @@ def __init__(
             )
 
 
+class SimpleReplicationClient(BaseHttpClient):
+    """No frills client for connecting to Replication endpoints.
+
+    Uses existing BaseHttpClient methods but replaces the 'agent' used to make the
+    request with one that supports HTTP and HTTPS.
+    Attributes:
+        agent: The custom Twisted Agent used for constructing the connection.
+    """
+
+    def __init__(
+        self,
+        hs: "HomeServer",
+    ):
+        """
+        Args:
+            hs: The HomeServer instance to pass in
+        """
+        super().__init__(hs)
+
+        # Use a pool, but a very small one.
+        pool = HTTPConnectionPool(self.reactor)
+        pool.maxPersistentPerHost = 5
+        pool.cachedConnectionTimeout = 2 * 60
+
+        self.agent: IAgent = ReplicationAgent(
+            hs.get_reactor(),
+            contextFactory=hs.get_http_client_context_factory(),
+            pool=pool,
+        )
```

Comment:
I don't think we need a custom class:

```suggestion
        self.agent: IAgent = Agent.usingEndpointFactory(
            reactor, ReplicationEndpointFactory(reactor, contextFactory)
        )
```

(I also don't think we need to use a custom `pool` and can use the default one?)

Reply:
I needed a custom class for the agent as the default `request()` function forces use of IDNA2008 in the 'hostname' area, which later will be in the way for passing the uri data in from the replication machinery when making unix sockets work. 
At this spot:
https://github.com/matrix-org/synapse/pull/15470/files#diff-209be8defd6cf72f8e6fd6c765ca7e11a172dc244c9f0dbcc57ece200fe53a2fR148-R159

I include the comment of why I need the overload.

The concept there will use the name of the worker as a 'hostname' of sorts to find the necessary connection details, and will work for both TCP and UNIX sockets. The hostnames will have to be less...strict than IDNA2008 allows for(specifically, need to allow underscores for Complement usage if nothing else). The alternative was a rather fumbly attempt at mutating the names into something compliant...but all this is for another PR. This was just laying the ground work. If you would rather I wait till then to put this in, I can.

Otherwise, I'm not married to the idea of a pool. Just seemed like low hanging fruit to save time making connections by reusing them instead of the teardown and reconstruct cost. For small homeservers, it would probably not see *much* use. But for m.org, I think it would be really useful. The default pool when none is provided is basically a one-use connection. It still creates it as a pool connection, but then is discarded when the connection is closed instead of saving it for reuse(if I understand that correctly)

Reply:
If we're going to do that, I think we should inherit from `Agent` so the changes are *much* more obvious; looking at the class it seemed to be fully copied & pasted to me.

I have mixed feelings on pre-emptively doing this vs. doing it in the PR which adds support for Unix, but these are all prep PRs, so I guess it makes sense? It is quite hard to see which changes will be used *after* this though so it is hard to reason about.

Reply:
Do you have it as branch at least so I can see the comparison?

Reply:
https://github.com/realtyem/synapse/compare/exp/unix-socket-repl-agent-alt...snapshot-diff-clokep

I believe the part you are looking for is in the commit titled `Wire up Unix sockets`